Source/DafnyCore/PrecedenceLinter.cs

```diff
@@ -0,0 +1,236 @@
+          // These expressions are handled below as being all-independent components, which means the
+          // formatting of their internal structure is not relevant to the enclosing context. Yet, the .tok of
+          // these expressions sits someone inside that internal structure (for example, the .tok of
+          // a SeqSelectExpr is the open-bracket). To avoid looking at that internal structure, we instead
+          // use the .StartToken for these expressions.
+          column = expr.StartToken.col;
+        } else {
+          column = expr.tok.col;
+        }
+        if (column < st.Column) {
+          st.Column = column;
+        }
+      }
+
+      // Our aim is to try to detect if some expressions are longer than the user intended.
+      // This can happen if the user accidentally left off parentheses around a lower-precedence
+      // operator (e.g., ==>). As a guide, we look at how the user formatted the code,
+      // that is, we inspect line and column information.
+
+      if (expr is BinaryExpr bin && (bin.Op == BinaryExpr.Opcode.Imp || bin.Op == BinaryExpr.Opcode.Exp || bin.Op == BinaryExpr.Opcode.Iff)) {
+        VisitLhsComponent(expr.tok, bin.E0,
+          // For
+          //   a)  LHS ==> RHS
+          //   b)  LHS ==>
+          // use LHS.StartToken as the left margin.
+          bin.E0.StartToken.line == expr.tok.line ? bin.E0.StartToken.col :
+          // For
+          //   c)  LHS0 &&
+          //       LHS1 ==> RHS
+          // use expr.tok (that is, the location of ==>) as the left margin. This is bound to generate a warning.
+          bin.E1.StartToken.line == expr.tok.line ? expr.tok.col :
+          // For
+          //   d)  LHS0 &&
+          //       LHS1 ==>
+          //   e)  LHS0 &&
+          //       LHS1
+          //       ==>
+          // use LHS.StartToken as the left margin.
+          bin.E0.StartToken.col,
+          "left-hand operand of " + BinaryExpr.OpcodeString(bin.Op));
+        VisitRhsComponent(expr.tok, bin.E1, "right-hand operand of " + BinaryExpr.OpcodeString(bin.Op));
+        return false; // indicate that we've already processed expr's subexpressions
+
+      } else if (expr is ITEExpr ifThenElse) {
+        VisitIndependentComponent(ifThenElse.Test);
+        VisitIndependentComponent(ifThenElse.Thn);
+        VisitRhsComponent(expr.tok, ifThenElse.Els, "else branch of if-then-else");
+        return false; // indicate that we've already processed expr's subexpressions
+
+      } else if (expr is QuantifierExpr quantifierExpr) {
+        Attributes.SubExpressions(quantifierExpr.Attributes).Iter(VisitIndependentComponent);
+        if (quantifierExpr.Range != null) {
+          VisitIndependentComponent(quantifierExpr.Range);
+        }
+        VisitRhsComponent(expr.tok, quantifierExpr.Term,
+          expr.tok.line == quantifierExpr.Term.StartToken.line ? expr.tok.col + 1 : quantifierExpr.Term.StartToken.col,
```

Comment:
Could you please ensure there is a test for the expression above, to show that it does not trigger errors? After that, I'm good approving it !

Reply:
I don't quite follow your example, because it is not legal Dafny. I added the following test case, which I hope captures the spirit of your example:

``` dafny
  datatype PQ = P(int) | Q(bool)

  method M(s: set<PQ>)
    requires
      (forall pq | pq in s :: match pq {
          case P(x) => true
          case Q(y) => y == false
      })
  {
  }
```

No precedence warning is generated for this example.